velox/functions/lib/TimeUtils.h

```diff
@@ -123,4 +126,218 @@ struct InitSessionTimezone {
+      return DateTimeUnit::kYear;
+    }
+  }
+  if (throwIfInvalid) {
+    VELOX_UNSUPPORTED("Unsupported datetime unit: {}", unitString);
+  }
+  return std::nullopt;
+}
+
+FOLLY_ALWAYS_INLINE void adjustDateTime(
+    std::tm& dateTime,
+    const DateTimeUnit& unit) {
+  switch (unit) {
+    case DateTimeUnit::kYear:
+      dateTime.tm_mon = 0;
+      dateTime.tm_yday = 0;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kQuarter:
+      dateTime.tm_mon = dateTime.tm_mon / 3 * 3;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kMonth:
+      dateTime.tm_mday = 1;
+      dateTime.tm_hour = 0;
+      dateTime.tm_min = 0;
+      dateTime.tm_sec = 0;
+      break;
+    case DateTimeUnit::kWeek:
+      // Subtract the truncation
+      dateTime.tm_mday -= dateTime.tm_wday == 0 ? 6 : dateTime.tm_wday - 1;
+      // Setting the day of the week to Monday
+      dateTime.tm_wday = 1;
+
+      // If the adjusted day of the month falls in the previous month
+      // Move to the previous month
+      if (dateTime.tm_mday < 1) {
+        dateTime.tm_mon -= 1;
+
+        // If the adjusted month falls in the previous year
+        // Set to December and Move to the previous year
+        if (dateTime.tm_mon < 0) {
+          dateTime.tm_mon = 11;
+          dateTime.tm_year -= 1;
+        }
+
+        // Calculate the correct day of the month based on the number of days
+        // in the adjusted month
+        static const int daysInMonth[] = {
+            31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+        int daysInPrevMonth = daysInMonth[dateTime.tm_mon];
+
+        // Adjust for leap year if February
+        if (dateTime.tm_mon == 1 && (dateTime.tm_year + 1900) % 4 == 0 &&
+            ((dateTime.tm_year + 1900) % 100 != 0 ||
+             (dateTime.tm_year + 1900) % 400 == 0)) {
+          daysInPrevMonth = 29;
+        }
+        // Set to the correct day in the previous month
+        dateTime.tm_mday += daysInPrevMonth;
+      }
+      dateTime.tm_hour = 0;
+      dateTime.tm_min = 0;
+      dateTime.tm_sec = 0;
+      break;
+    case DateTimeUnit::kDay:
+      dateTime.tm_hour = 0;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kHour:
+      dateTime.tm_min = 0;
+      FMT_FALLTHROUGH;
+    case DateTimeUnit::kMinute:
+      dateTime.tm_sec = 0;
+      break;
+    default:
+      VELOX_UNREACHABLE();
+  }
+}
+
+/// For fixed interval like second, minute, hour, day and week
+/// we can truncate date by a simple arithmetic expression:
+/// floor(seconds / intervalSeconds) * intervalSeconds.
+FOLLY_ALWAYS_INLINE Timestamp
+adjustEpoch(int64_t seconds, int64_t intervalSeconds) {
+  int64_t s = seconds / intervalSeconds;
+  if (seconds < 0 && seconds % intervalSeconds) {
+    s = s - 1;
+  }
+  int64_t truncedSeconds = s * intervalSeconds;
+  return Timestamp(truncedSeconds, 0);
+}
+
+// Returns timestamp truncated to the unit specified by the format.
+FOLLY_ALWAYS_INLINE Timestamp dateTrunc(
+    const DateTimeUnit unit,
```

Comment:
Don't add the const to enum class, you could consider it as integer like.
https://stackoverflow.com/questions/44749658/c-is-it-better-to-pass-an-enum-class-as-value-or-const-reference

Reply:
@jinchengchenghh Thanks for review. It seems to be used this way in velox, https://github.com/facebookincubator/velox/blob/1e7a82f0379da727412002d8192758ef90de95b8/velox/functions/prestosql/DateTimeFunctions.h#L929

Reply:
No need to add const when passing by value.